### PR TITLE
Fullscreen for client

### PR DESCRIPTION
**Pull request by akirmse**

### CODE REVIEW

File: clientd3d/intrface.h [hunk 1]
```diff
@@ -40,9 +40,8 @@
 #define GRAPHICS_TOP_HEIGHT 24
 
 /* Inventory area */
-//#define INVENTORY_MIN_WIDTH (170 + LEFT_BORDER)
-#define INVENTORY_MIN_WIDTH (185 + LEFT_BORDER + 67)
-#define INVENTORY_MAX_WIDTH (250 + LEFT_BORDER)
+#define INVENTORY_MIN_WIDTH (215 + LEFT_BORDER + 67)
+#define INVENTORY_MAX_WIDTH 9000
```
Comment (akirmse) [inline clientd3d/intrface.h:44]: Is this effectively infinite?  If so, maybe just remove whatever code is checking it?

File: clientd3d/d3drender.c [hunk 2]
```diff
@@ -13,8 +13,9 @@
 #define	TEX_CACHE_MAX_WALLMASK	2000000
 #define	TEX_CACHE_MAX_EFFECT	2000000
 #define	TEX_CACHE_MAX_PARTICLE	1000000
-#define FOV_H					((gD3DRect.right - gD3DRect.left) / (float)(MAXX * stretchfactor) * (-PI / 3.6f))
-#define FOV_V					((gD3DRect.bottom - gD3DRect.top) / (float)(MAXY * stretchfactor) * (PI / 6.0f))
+// Define field of views with magic numbers for tuning
+#define FOV_H					((gD3DRect.right - gD3DRect.left) / (float)(main_viewport_width) * (-PI/3.6f))
```
Comment (akirmse) [inline clientd3d/d3drender.c:17]: Make this an inline function instead?  Macros are gross.

File: clientd3d/d3drender.c [hunk 3]
```diff
@@ -6896,6 +6891,9 @@ void D3DRenderObjectsDraw(d3d_render_pool_new *pPool, room_type *room,
 			bottomRight.z = 0;
 			bottomRight.w = 1.0f;
 
+
+         //stretchfactor = config.viewport_stretchfactor;
```
Comment (akirmse) [inline clientd3d/d3drender.c:6895]: Delete this.  I think you can actually delete all mentions of stretchfactor in this file?

File: clientd3d/d3drender.c [hunk 4]
```diff
@@ -7951,8 +7950,8 @@ void D3DRenderPlayerOverlaysDraw(d3d_render_pool_new *pPool, room_type *room, Dr
 	d3d_render_packet_new	*pPacket;
 	d3d_render_chunk_new	*pChunk;
 
-	screenW = (float)(gD3DRect.right - gD3DRect.left) / (float)(MAXX * stretchfactor);
-	screenH = (float)(gD3DRect.bottom - gD3DRect.top) / (float)(MAXY * stretchfactor);
+   screenW = /*(float)(gD3DRect.right - gD3DRect.left) /*/ (float)(main_viewport_width);
```
Comment (akirmse) [inline clientd3d/d3drender.c:7953]: Delete dead code.

File: clientd3d/d3drender.c [hunk 5]
```diff
@@ -8136,8 +8136,8 @@ void D3DRenderPlayerOverlayOverlaysDraw(d3d_render_pool_new *pPool, list_type ov
 	d3d_render_packet_new	*pPacket;
 	d3d_render_chunk_new	*pChunk;
 
-	screenW = (float)(gD3DRect.right - gD3DRect.left) / (float)(MAXX * (stretchfactor * 1.5f));
-	screenH = (float)(gD3DRect.bottom - gD3DRect.top) / (float)(MAXY * (stretchfactor * 1.5f));
+   screenW = (float)(gD3DRect.right - gD3DRect.left) / (float)main_viewport_width;// (MAXX * (stretchfactor * 1.5f));
```
Comment (akirmse) [inline clientd3d/d3drender.c:8139]: Delete dead code.

**Review by akirmse — COMMENTED** [review]
One overall comment: is the software rendering mode still alive?  Does it still work?

File: clientd3d/config.c [hunk 6]
```diff
@@ -82,8 +82,8 @@ static char INIRight[]       = "NormalRight";
 static char INITop[]         = "NormalTop";
 static char INIBottom[]      = "NormalBottom";
 static char INIShow[]        = "Show";
-static char INIMaxX[]        = "MaxX";
-static char INIMaxY[]        = "MaxY";
+static char INIMAXX[]        = "MAXX";
```
[the remark this answers is not in the document]
Comment (sicoderealm) [reply clientd3d/config.c]: This should now be reverted. Originally I planned to change the variable name backing this but ended up reverting - however I missed this.

File: clientd3d/client.h [hunk 7]
```diff
@@ -77,6 +77,10 @@ extern void ClearMessageQueue(void);
 
 extern Bool is_foreground;   // True when program is in the foreground
 
+// Main client windows current viewport area
+extern int main_viewport_width;
```
[the remark this answers is not in the document]
Comment (sicoderealm) [reply clientd3d/client.h]: good call

File: clientd3d/draw3d.c [hunk 8]
```diff
@@ -333,17 +333,20 @@ void DrawViewTreatment()
...
 
 	for( i = iOffset; i < iOffset + ( NUM_VIEW_ELEMENTS / 2 ); i++ )
 	{
-		BitCopy( pBitsTarget, iWidthTarget, ViewElements[i].x, ViewElements[i].y, ViewElements[i].width, ViewElements[i].height,
-					ViewElements[i].bits, 0, 0, DIBWIDTH( ViewElements[i].width ), OBB_FLIP | OBB_TRANSPARENT );
+		// The classic client has some visible artifacts when opening doors and changing zones relating to the corner graphics. This is more noticable on a larger viewport
+		// The following BitCopy causes strange copies of corners all over the place during this code flow.
+		//BitCopy( pBitsTarget, iWidthTarget, ViewElements[i].x, ViewElements[i].y, ViewElements[i].width, ViewElements[i].height, 
+		// ViewElements[i].bits, 0, 0, DIBWIDTH( ViewElements[i].width ), OBB_FLIP | OBB_TRANSPARENT );
+
```
[the remark this answers is not in the document]
Comment (sicoderealm) [reply clientd3d/draw3d.c]: purged!

File: clientd3d/drawdefs.h [hunk 9]
```diff
@@ -26,9 +26,11 @@
...
 #define TURN_BACK(angle) ((angle + NUMDEGREES / 2) % NUMDEGREES)
 
-/* Max viewport size -- must be DWORD aligned for WinG */
-#define MAXX 452
-#define MAXY 276
+/* Original classic view port size -- must be DWORD aligned for WinG */
+#define MAXX 1452
+#define MAXY 1276
+#define MAXYX_PROPTION 1.68
+#define MAXXY_PROPTION 0.59
```
[the remark this answers is not in the document]
Comment (sicoderealm) [reply clientd3d/drawdefs.h]: updated to include "aspect ratio" in the names here and added reciprocal definition

File: clientd3d/intrface.h (hunk 1 above)
Comment (sicoderealm) [reply clientd3d/intrface.h]: I've removed as much of this now as I can. The minimap uses this max value to define a bitmap for the maximum size at the start. I've kept a max minimap value to support this - please let me know if there is a better alternative for that last constant

File: clientd3d/d3drender.c (hunk 2 above)
Comment (sicoderealm) [reply clientd3d/d3drender.c]: new inline functions added

File: clientd3d/d3drender.c (hunk 3 above)
Comment (sicoderealm) [reply clientd3d/d3drender.c]: done

File: clientd3d/d3drender.c (hunk 5 above)
Comment (sicoderealm) [reply clientd3d/d3drender.c]: done

File: clientd3d/draw3d.c [hunk 10]
```diff
@@ -327,34 +333,33 @@ void DrawPostOverlayEffects(room_type* room, Draw3DParams* params)
...
+      {
+         BitCopy(pBitsTarget, iWidthTarget, ViewElements[i].x, ViewElements[i].y, ViewElements[i].width, ViewElements[i].height,
+            ViewElements[i].bits, 0, 0, DIBWIDTH(ViewElements[i].width), OBB_FLIP | OBB_TRANSPARENT);
+         GdiFlush();
+      }
+   }
+
+   //	Ensure that border, which covers up parts of view treatment, is drawn.
+   DrawGridBorder();
+   GdiFlush();
```
Comment (akirmse) [inline clientd3d/draw3d.c:358]: Any idea why there are 5 GdiFlush calls here?  I know they were already here, but it's weird.

File: clientd3d/intrface.h [hunk 11]
```diff
@@ -40,13 +40,10 @@
...
-//#define INVENTORY_MIN_WIDTH (170 + LEFT_BORDER)
-#define INVENTORY_MIN_WIDTH (185 + LEFT_BORDER + 67)
-#define INVENTORY_MAX_WIDTH (250 + LEFT_BORDER)
+#define INVENTORY_MIN_WIDTH (215 + LEFT_BORDER + 67)
 
-//	MiniMap area.
-#define MINIMAP_MAX_WIDTH	( INVENTORY_MAX_WIDTH + 3 ) & ~3
-#define MINIMAP_MAX_HEIGHT	MINIMAP_MAX_WIDTH
+/* MiniMap area */
+#define MINIMAP_MAX	9000
```
Comment (akirmse) [inline clientd3d/intrface.h:46]: Name could be improved--MINIMAP_MAX_AREA?

File: clientd3d/draw3d.c [hunk 12]
```diff
@@ -846,7 +851,7 @@ void DrawMiniMap(room_type *room, Draw3DParams *params)
    CopyCurrentAreaMiniMap(&areaMiniMap);
    area = areaMiniMap;
    area.x = area.y = 0;
-   MapDraw(gMiniMapDC, gMiniMapBits, &area, room, MINIMAP_MAX_WIDTH, TRUE); // cx = MINIMAP_MAX_WIDTH
+   MapDraw(gMiniMapDC, gMiniMapBits, &area, room, MINIMAP_MAX, TRUE); // cx = MINIMAP_MAX_WIDTH
```
Comment (akirmse) [inline clientd3d/draw3d.c:854]: Delete dead code at end of line.

**Review by akirmse — COMMENTED** [review]
Looking pretty good!

The one bug I see is that in software mode, if I click the window's maximize button, I get a crash.  Let me know if you can't repro, and I'll debug.

File: clientd3d/draw3d.c (hunk 10 above)
Comment (sicoderealm) [reply clientd3d/draw3d.c:358]: I have no knowledge on why there are 5 calls here but I've just removed these and see no obvious issues in doing so

File: clientd3d/intrface.h (hunk 11 above)
Comment (sicoderealm) [reply clientd3d/intrface.h:46]: applied

File: clientd3d/draw3d.c (hunk 12 above)
Comment (sicoderealm) [reply clientd3d/draw3d.c]: done

**Comment by davidklaw** [comment]
@sicoderealm 👏

**Review by sicoderealm — APPROVED** [review]

**Comment by davidklaw** [comment]
I haven't taken a look at this yet, but given @akirmse's merging I'd just like to thank @sicoderealm for coming back to make this happen. And @akirmse as well, of course. I'm looking forward to seeing this released!

**Pull request merged**